### PR TITLE
copy_(Sparse, Sparse) for sparse tensor

### LINKED ISSUE #8330: x.copy_(y) doesn't work with sparse x
```
>>> i = torch.LongTensor([[0, 2], [1, 0], [1, 2]])
>>> v = torch.FloatTensor([3,      4,      5    ])
>>> x = torch.sparse.FloatTensor(i.t(), v, torch.Size([2,3]))
>>> x2 = torch.sparse.FloatTensor(i.t(), v, torch.Size([2,3]))
>>> x.copy_(x2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: copy does not support torch.sparse.FloatTensor to torch.sparse.FloatTensor copy.
```

There isn't really any reason this shouldn't work; internally we have all the pieces implemented already.

### PR DESCRIPTION
**Pull request by weiyangfb**
## summary
- fix #8330
- add `torch.copy_(Sparse, Sparse)` with autograd support

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -754,16 +754,12 @@
   dispatch:
     CPU: _log1p__cpu
     CUDA: _log1p__cuda
-    SparseCPU: log1p_sparse_
```
Comment (SsnL) [inline aten/src/ATen/native/native_functions.yaml]: why are these removed? O:

Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: oops, that's strange... I will revert it

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1772,8 +1774,7 @@
     SparseCPU: hspmm_sparse_cpu
     SparseCUDA: hspmm_sparse_cuda
 
-# This "raw copy" doesn't handle conversions NOR does it handle non-blocking.
-- func: raw_copy_sparse_(Tensor self, Tensor src) -> Tensor
+- func: raw_copy_sparse_(Tensor self, Tensor src, bool non_blocking=false) -> Tensor
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: OK, so what about conversions? If I copy from FloatTensor to DoubleTensor a conversion happens. What happens here?

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: Also, if all of this is fixed, let's call it something besides `raw_copy_sparse`, like `copy_sparse_to_sparse`.

Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: So what is the expected output after type conversions here? For instance, in dense tensor, when copy a FloatTensor to DoubleTensor, all values in FloatTensor will cast Float to Double. 
```
x = torch.tensor([1., 1.], dtype=torch.float32)
y = torch.tensor([2., 2.], dtype=torch.float64)
y.copy_(x)
print(y, y.dtype)
---------------
(tensor([1., 1.], dtype=torch.float64), torch.float64)
```
And this also holds for sparse tensor. `raw_copy_sparse_ ` should have already served this purpose I guess?

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: Yes, so maybe this already handles conversions now that it was changed from `clone()` to `copy()`. In that case, nothing more to do but change the function name :)

Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: Right, now sparse copy_() does the type conversion by using [dense copy() for indices and values](https://github.com/weiyangfb/pytorch/blob/83e626d53e4340807e4649de2e3e06a0b0dfaf92/aten/src/ATen/native/sparse/SparseUtils.h#L57-L59)

**Comment by weiyangfb** [comment]
@ezyang can I get a stamp on this?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1425,6 +1425,8 @@
 
 - func: zero_(Tensor self) -> Tensor
 
+
+
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml]: nit: remove whitespace

File: test/test_sparse.py [hunk 4]
```diff
@@ -339,6 +339,45 @@ def test_clone(self):
...
 
+    def test_S_to_S_copy_(self):
+        expected_output = torch.tensor([3., 4., 5.])
+        copy_src = torch.sparse.FloatTensor(
+            torch.LongTensor([[0], [1], [2]]).t(),
+            torch.FloatTensor([3, 4, 5]),
+            torch.Size([3]))
+
+        if self.is_cuda:
+            input = torch.cuda.sparse.FloatTensor(
```
Comment (zou3519) [inline test/test_sparse.py]: nit: maybe use the new torch.sparse_coo_tensor(...) constructors?

**Comment by SsnL** [comment]
does this support autograd similarly to dense tensors? E.g.,
```py
>>> x = torch.randn(3)
>>> y = torch.randn(3, requires_grad=True)
>>> x.copy_(y).sum().backward()
>>> y.grad
tensor([1., 1., 1.])

```

File: test/test_sparse.py [hunk 5]
```diff
@@ -339,6 +339,45 @@ def test_clone(self):
...
+        input_dtype = input.dtype
+        copy_src = torch.sparse.DoubleTensor(
+            torch.LongTensor([[0], [1], [2]]).t(),
+            torch.DoubleTensor([3, 4, 5]),
+            torch.Size([3]))
+        input.copy_(copy_src, non_blocking=True)
+        self.assertEqual(input_dtype, input.dtype)
+        self.assertEqual(expected_output, input.to_dense())
+
+        # test no broadcast
```
Comment (zou3519) [inline test/test_sparse.py]: Are you testing that copy broadcasts?

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -1772,8 +1774,7 @@
     SparseCPU: hspmm_sparse_cpu
     SparseCUDA: hspmm_sparse_cuda
 
-# This "raw copy" doesn't handle conversions NOR does it handle non-blocking.
-- func: raw_copy_sparse_(Tensor self, Tensor src) -> Tensor
+- func: copy_sparse_to_sparse_(Tensor self, Tensor src, bool non_blocking=false) -> Tensor
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml]: I'm confused -- where is the implementation for this?

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: I thought the extra whitespace at couple different places in this file were used to create "boundaries". Should I follow the same convention?

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: It's here: https://github.com/weiyangfb/pytorch/blob/90dea26d6903c401280b443be9f33a36da2d27c1/aten/src/ATen/native/sparse/SparseTensor.cpp#L247-L255

File: test/test_sparse.py (hunk 5 above)
Comment (weiyangfb) [reply test/test_sparse.py]: I think I was assuring copy_() is not broadcast-able in sparse, as opposed to in dense.

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: wait, I think not, but I define something like: https://github.com/weiyangfb/pytorch/blob/90dea26d6903c401280b443be9f33a36da2d27c1/aten/src/ATen/native/native_functions.yaml#L1777-L1781

maybe it should just work?

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: I think that's a valid point, extra space might not be a good convention to separate different functions, either does it applied at all places in the file. Maybe to use comments is a better way to do it...

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml]: Gotcha, sorry for missing that line. Yes that would do it

**Comment by weiyangfb** [comment]
@SsnL Good question! The answer is no, I haven't defined the backward for copy_() in sparse yet...

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 7]
```diff
@@ -244,11 +244,11 @@ Tensor sparse_to_dense(const SparseTensor& self) {
   return dst.add_(self);
 }
 
-SparseTensor& copy_sparse_(SparseTensor& self, const SparseTensor& src) {
+SparseTensor& copy_sparse_(SparseTensor& self, const SparseTensor& src, bool non_blocking) {
```
Comment (SsnL) [inline aten/src/ATen/native/sparse/SparseTensor.cpp]: Hmm does this work if `src` and `self` are on different cuda devices?

**Review by ssnl — COMMENTED** [review]
I have some concerns when `src` and `self` have same backend but are on different devices. Commented inline.

File: aten/src/ATen/native/sparse/SparseTensor.cpp (hunk 7 above)
Comment (weiyangfb) [reply aten/src/ATen/native/sparse/SparseTensor.cpp]: Yes, this works:
```
i = torch.LongTensor([[0, 2], [1, 0], [1, 2]]).t()
v = torch.FloatTensor([3, 4, 5])
size = torch.Size([2,3])
x = torch.sparse_coo_tensor(i, torch.ones_like(v), size, device=torch.device('cuda:0'))
x2 = torch.sparse_coo_tensor(i, v, size, requires_grad=True, device=torch.device('cuda:1'))
x.copy_(x2)
#####################
torch.cuda.sparse.FloatTensor of size (2,3) with indices:
tensor([[0, 1, 1],
        [2, 0, 2]], device='cuda:0', grad_fn=<Error>)
and values:
tensor([3., 4., 5.], device='cuda:0', grad_fn=<Error>)
```

File: tools/autograd/templates/Functions.cpp [hunk 8]
```diff
@@ -1872,6 +1872,15 @@ Tensor log1p_backward(const Tensor& grad, const Tensor& self) {
   return grad / (self + 1);
 }
 
+std::tuple<Tensor, Tensor> copy_sparse_to_sparse_backward(const Tensor& grad, const Tensor& self, const Tensor& src, bool non_blocking) {
+  auto self_grad = at::native::sparse_coo_tensor( self._indices().type().tensor({0}), self._values().type().tensor({0}), self.sizes());
+  if (src.is_cuda() != grad.is_cuda()) { // cuda -> cpu or cpu -> cuda
```
[the remark this answers is not in the document]
Comment (weiyangfb) [reply tools/autograd/templates/Functions.cpp]: I made changes in the PR to reuse `CopyBackward ` and `s_copy_ ` for sparse `copy_`

**Comment by weiyangfb** [comment]
@SsnL feel free to take a look when you have time :)

File: aten/src/ATen/templates/TypeDefault.cpp [hunk 9]
```diff
@@ -28,19 +29,11 @@ Tensor TypeDefault::copy(const Tensor & src, bool non_blocking, optional<Device>
...
-    auto indices_copy = this_dense_idx.copy(indices, non_blocking);
-    auto values_copy = this_dense.copy(values, non_blocking);
-    return _sparse_coo_tensor_unsafe(indices_copy, values_copy, src.sizes());
-  } else {
-    Tensor r = this->tensor(src.sizes());
-    r.copy_(src, non_blocking);
-    return r;
-  }
+  Tensor r;
+  if (is_sparse()) r = this->native_tensor();
```
Comment (SsnL) [inline aten/src/ATen/templates/TypeDefault.cpp]: Interesting. What does `native_tensor` do?

File: tools/autograd/templates/VariableType.cpp [hunk 10]
```diff
@@ -416,7 +416,9 @@ Tensor & VariableType::s_copy_(Tensor & self, const Tensor & src, bool non_block
       grad_fn->src_device = src.get_device();
     }
   }
-  baseType->s_copy_(self_, src_, non_blocking);
+  if (self.is_sparse() && src.is_sparse()) baseType->copy_sparse_to_sparse_(self_, src_, non_blocking);
+  else if (!self.is_sparse() && !src.is_sparse()) baseType->s_copy_(self_, src_, non_blocking);
+  else AT_ERROR("s_copy_() between dense and sparse Tensors is not implemented! Found self type = ", self.type(), " and src type = ", src.type());
```
Comment (SsnL) [inline tools/autograd/templates/VariableType.cpp]: Probably should just say `copy()` since `s_copy` is non-public.

File: aten/src/ATen/native/native_functions.yaml [hunk 11]
```diff
@@ -2068,8 +2068,8 @@
...
```
Comment (SsnL) [inline aten/src/ATen/native/native_functions.yaml]: Is this function exposed in python? If so, could you blacklist it in `gen_python_functions.py`?

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 12]
```diff
@@ -204,7 +204,7 @@ SparseTensor new_with_tensor_and_size_sparse(const LongTensor& indices, const Te
 
 SparseTensor clone_sparse(const SparseTensor& self) {
   SparseTensor other = new_with_dims_and_size_sparse(self.type(), self._sparseDims(), self._denseDims(), self.sizes());
-  _copy_into_sparse(other, _get_sparse_impl(self)->indices(), _get_sparse_impl(self)->values());
+  _copy_into_sparse(other, _get_sparse_impl(self)->indices(), _get_sparse_impl(self)->values(), false);
```
Comment (SsnL) [inline aten/src/ATen/native/sparse/SparseTensor.cpp]: So if i understand correctly, for CUDA types, it is okay if `.clone()` is async (and in fact is preferred); for CPU types, there is no such thing as async copy. And `clone()` doesn't do cross device copy.... maybe we should just set to `true` here? Correct me if I'm wrong though.

**Review by ssnl — COMMENTED** [review]
Looks good overall. But I have some questions and suggestions.

File: aten/src/ATen/templates/TypeDefault.cpp (hunk 9 above)
Comment (weiyangfb) [reply aten/src/ATen/templates/TypeDefault.cpp]: it creates a new sparse tensor given `Type` passed in: https://github.com/pytorch/pytorch/blob/e7e10e60e09c8eb133b1addab1490b6320a7c366/aten/src/ATen/native/native_functions.yaml#L1899-L1903

File: aten/src/ATen/native/native_functions.yaml (hunk 11 above)
Comment (weiyangfb) [reply aten/src/ATen/native/native_functions.yaml]: right, it is not exposed to python, I will add it to blacklist at `gen_python_functions.py `

File: aten/src/ATen/native/sparse/SparseTensor.cpp (hunk 12 above)
Comment (weiyangfb) [reply aten/src/ATen/native/sparse/SparseTensor.cpp]: Yeah, sure, I agree that set `non_blocking=true` here might make more sense. So IIUC, `non_blocking=true` should not cause any issue when:
```
x = SparseTensor.clone()
y = x.clone()
```
Is it correct?

**Review by ssnl — APPROVED** [review]
